arch/risc-v/src/bl602/bl602_dma.c

```diff
@@ -0,0 +1,491 @@
+/****************************************************************************
+ * arch/risc-v/src/bl602/bl602_dma.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/arch.h>
+
+#include <stdint.h>
+#include <errno.h>
+#include <assert.h>
+#include <debug.h>
+
+#include "chip.h"
+#include "riscv_internal.h"
+#include "hardware/bl602_dma.h"
+#include "bl602_dma.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/* This structure describes one DMA channel */
+
+struct dma_channel_s
+{
+  uint8_t              chan;          /* DMA channel number (0-BL602_DMA_NCHANNELS) */
+  bool                 inuse;         /* TRUE: The DMA channel is in use */
+  bl602_dma_callback_t callback;      /* Callback invoked when the DMA completes */
+  void                 *arg;          /* Argument passed to callback function */
+};
+
+/* This structure describes the state of the DMA controller */
+
+struct dma_controller_s
+{
+  sem_t exclsem; /* Protects channel table */
```

Comment:
Let's align with https://github.com/apache/incubator-nuttx/pull/6965 and start using nxmutex here.

Reply:
I was not sure if that PR was a prerequisite for the conversion. If the machinery already exists I can use it.

Reply:
Ok. Let's update that PR after this one is merged.

Reply:
The machinery is already there without #6965:
https://github.com/apache/incubator-nuttx/blob/master/include/nuttx/mutex.h

Reply:
Cool I'll update this to use it.

Reply:
Fixed in the new revision of #6965.